src/ui/EditorNS/languagecache.cpp

```diff
@@ -0,0 +1,139 @@
+#include <QFileInfo>
+#include <QJsonDocument>
+#include <QJsonObject>
+#include <QJsonValue>
+
+#include "include/EditorNS/languagecache.h"
+#include "include/notepadqq.h"
+
+namespace EditorNS {
+
+Language::Language(const Language& o) :
+    id(o.id),
+    name(o.name),
+    mime(o.mime),
+    mode(o.mode),
+    fileNames(o.fileNames),
+    fileExtensions(o.fileExtensions),
+    firstNonBlankLine(o.firstNonBlankLine) 
+{
+}
+
+Language::Language(Language&& o) noexcept :
+    id(std::move(o.id)),
+    name(std::move(o.name)),
+    mime(std::move(o.mime)),
+    mode(std::move(o.mode)),
+    fileNames(std::move(o.fileNames)),
+    fileExtensions(std::move(o.fileExtensions)),
+    firstNonBlankLine(std::move(o.firstNonBlankLine))
+{
+}
+
+Language& Language::operator=(const Language& o)
+{
+    id = o.id;
+    name = o.name;
+    mime = o.mime;
+    mode = o.mode;
+    fileNames = o.fileNames;
+    fileExtensions = o.fileExtensions;
+    firstNonBlankLine = o.firstNonBlankLine;
+    return *this;
+}
+
+Language& Language::operator=(Language&& o)
+{
+    id = std::move(o.id);
+    name = std::move(o.name);
+    mime = std::move(o.mime);
+    mode = std::move(o.mode);
+    fileNames = std::move(o.fileNames);
+    fileExtensions = std::move(o.fileExtensions);
+    firstNonBlankLine = std::move(o.firstNonBlankLine);
+    return *this;
+}
+
+LanguageCache::LanguageCache()
+{
+    QFileInfo fileInfo(Notepadqq::editorPath());
+    QString fileName = fileInfo.absolutePath() + "/Languages.json";
+    QFile scriptFile(fileName);
+    scriptFile.open(QIODevice::ReadOnly | QIODevice::Text);
+    QJsonDocument json = QJsonDocument::fromJson(scriptFile.readAll());
+    scriptFile.close();
+
+    // Reserve the space we need to prevent resize and copy constructs.
+    m_languages.reserve(json.object().count());
+
+    // Begin iterating our QJsonDocument's object and adding languages.
+    for (const auto& key : json.object().keys()) {
+        auto mode = json.object().value(key).toObject();
+        Language newMode;
+        newMode.id = key;
+        newMode.name = mode.value("name").toString();
+        newMode.mime = mode.value("mime").toString();
+        newMode.mode = mode.value("mode").toString();
+        newMode.fileNames = mode.value("fileNames")
+            .toVariant().toStringList();
+        newMode.fileExtensions = mode.value("fileExtensions")
+            .toVariant().toStringList();
+        newMode.firstNonBlankLine = mode.value("firstNonBlankLine")
+            .toVariant().toStringList();
+        m_languages.append(std::move(newMode));
+    }
+}
+
+const LanguageList& LanguageCache::languages()
+{
+    return m_languages;
+}
+
+int LanguageCache::lookupById(const QString& id)
+{
+    auto end = m_languages.constEnd();
+    for (auto it = m_languages.constBegin(); it != end; ++ it) {
```

Comment:
This loops (and the following) should probably be using `std::find_if`.

Reply:
```
int LanguageCache::lookupById(const QString& id)
{
    auto it = std::find_if (m_languages.constBegin(), m_languages.constEnd(),
        [&id] (const Language& l) {
            return (l.id == id);
        });
    if (it != m_languages.constEnd())
        return it - m_languages.constEnd();
    return -1;
}
```

Something like this?